Nodejs/Product/TestAdapter/PackageJsonTestDiscoverer.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.NodejsTools.Npm;
+using Microsoft.NodejsTools.TestAdapter.TestFrameworks;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+
+namespace Microsoft.NodejsTools.TestAdapter
+{
+    [FileExtension(".json")]
+    [DefaultExecutorUri(NodejsConstants.PackageJsonExecutorUriString)]
+    public sealed class PackageJsonTestDiscoverer : ITestDiscoverer
+    {
+        public void DiscoverTests(IEnumerable<string> sources, IDiscoveryContext discoveryContext, IMessageLogger logger, ITestCaseDiscoverySink discoverySink)
+        {
+            foreach (var source in sources)
+            {
+                // we're only interested in package.json files here.
+                if (PackageJsonFactory.IsPackageJsonFile(source))
+                {
+                    this.DiscoverTestFiles(source, logger, discoverySink);
+                }
+            }
+        }
+
+        private void DiscoverTestFiles(string packageJsonPath, IMessageLogger logger, ITestCaseDiscoverySink discoverySink)
+        {
+            logger.SendMessage(TestMessageLevel.Informational, $"Parsing '{packageJsonPath}'.");
+
+            var packageJson = PackageJsonFactory.Create(packageJsonPath);
+            if (string.IsNullOrEmpty(packageJson.TestRoot))
+            {
+                logger.SendMessage(TestMessageLevel.Informational, "No vsTestOptions|testRoot specified.");
+                return;
+            }
+
+            var workingDir = Path.GetDirectoryName(packageJsonPath);
+            var testFolderPath = Path.Combine(workingDir, packageJson.TestRoot);
+            TestFramework testFx = null;
+
+            foreach (var dep in packageJson.AllDependencies)
+            {
+                testFx = FrameworkDiscover.Intance.Get(dep.Name);
+                if (testFx != null)
+                {
+                    break;
+                }
+            }
+            testFx = testFx ?? FrameworkDiscover.Intance.Get("ExportRunner");
+
+            var nodeExePath = Nodejs.GetPathToNodeExecutableFromEnvironment();
+
+            if (!File.Exists(nodeExePath))
+            {
+                logger.SendMessage(TestMessageLevel.Error, "Node.exe was not found. Please install Node.js before running tests.");
+                return;
+            }
+
+            var fileList = Directory.GetFiles(testFolderPath, "*.js", SearchOption.AllDirectories);
+            var files = string.Join(";", fileList);
```

Comment:
>var files = string.Join(";", fileList); [](start = 12, length = 39)

This seems likely to be expensive.  Consider checking the logging level before evaluating it.

Reply:
All log levels are always outputted, they only change the color of the output.